README.md
# livereload-base

This is a [LiveReload][livereload] server implementation that exposes an API for
communicating with LiveReload-compatible clients. The watcher implementation is
intentionally excluded to allow developers to bring their own without needing to
worry about implementing the LiveReload-compatible server.

## Example

```js
const livereload = require('livereload-base')

// create and start the server
const server = livereload.createServer()
server.listen(() => console.log('livereload server listening'))

// bring your own watcher...
watcher.on('change', (file) => server.reload(file.path))
```

## API

This is a very new module, so I expect the API to change fairly frequently as
real-world usage occurs.

### createServer()

Creates a new `Server` instance. There is no configuration available at the
time, but that may likely change in the future.

### Server()

This base class represents a [LiveReload][livereload]-compatible server, which
includes both HTTP and Websockets as described in the LiveReload
[protocol specification][livereload-protocol]

### Server#listen(callback)

Starts the server, which will listen on port `35729`, as defined by the
[spec][livereload-protocol].

The `callback` is optional, and will be invoked for the `listening` event, which
indicates that the server is ready to accept connections.

### Server#reload(file)

Broadcasts a `reload` command to all the connected clients.

The `file` parameter tells the client which file was changed. An absolute path
is preferred, but a simple basename will suffice.

### Server#alert(message)

Broadcasts an `alert` command to all connected clients.

The `message` parameter will be sent to the clients, which may be communicated
to the end-user.


[livereload]: http://livereload.com/
[livereload-protocol]: http://feedback.livereload.com/knowledgebase/articles/86174-livereload-protocol


HISTORY.md

0.1.0 / 2016-10-23
==================

  * test: add test suite

0.0.0 / 2015-10-20
==================

:sparkles:


test/index.js
/* eslint-env mocha */

'use strict'

const assert = require('chai').assert
const fs = require('fs')
const livereload = require('..')
const request = require('supertest')
const Server = require('../lib/websocket')
const WebSocket = require('ws')

const host = 'localhost:35729'
const js = fs.readFileSync(require.resolve('livereload-js/dist/livereload.js'))

describe('livereload-base', function () {
  describe('.Server', function () {
    it('should export the Server instance directly', function () {
      assert.strictEqual(livereload.Server, Server)
    })
  })

  describe('.createServer()', function () {
    const server = livereload.createServer()

    before(function (done) {
      server.listen(done)
    })

    afterEach(function (done) {
      for (const client of server.wss.clients) client.terminate()
      setTimeout(done, 100)
    })

    after(function (done) {
      server.close(done)
    })

    it('should return a Server instance', function () {
      assert.instanceOf(server, Server)
    })

    context('http', function () {
      const base = `http://${host}`

      it('should serve the livereload client javascript', function (done) {
        request(base)
          .get('/livereload.js')
          .expect('Content-Type', 'text/javascript')
          .expect(200, js.toString(), done)
      })

      it('should serve a 404 for other routes', function (done) {
        request(base)
          .get('/')
          .expect(404, done)
      })

      it('should serve a 404 for favicon', function (done) {
        request(base)
          .get('/favicon.ico')
          .expect(404, done)
      })
    })

    context('websocket', function () {
      const base = `ws://${host}/livereload`

      it('should start a websocket server', function (done) {
        const ws = new WebSocket(base)
        ws.on('close', () => done())
        ws.on('error', (err) => done(err))
        ws.on('open', () => ws.close())
      })

      context('on client hello', function () {
        it('should respond with another hello', function (done) {
          const ws = new WebSocket(base)
          ws.on('close', () => done())
          ws.on('error', (err) => done(err))
          ws.on('open', () => ws.send(JSON.stringify({ command: 'hello' })))
          ws.on('message', (payload) => {
            const message = JSON.parse(payload)
            assert.equal(message.command, 'hello')
            assert.equal(message.serverName, 'livereload-base')
            assert.isArray(message.protocols)
            ws.close()
          })
        })
      })

      context('on client *', function () {
        it('should be a no-op', function (done) {
          const ws = new WebSocket(base)
          ws.on('close', () => done())
          ws.on('error', (err) => done(err))
          ws.on('message', (message) => done(new Error('no response expected')))
          ws.on('open', () => {
            const payload = JSON.stringify({ command: 'something' })
            ws.send(payload, function () {
              // delay long enough to see if server responds
              setTimeout(() => ws.close(), 100)
            })
          })
        })
      })

      context('on server reload', function () {
        it('should send a reload command', function (done) {
          const ws = new WebSocket(base)
          ws.on('close', () => done())
          ws.on('error', (err) => done(err))
          ws.on('message', (payload) => {
            const message = JSON.parse(payload)
            assert.equal(message.command, 'reload')
            assert.equal(message.path, 'index.js')
            assert.isTrue(message.liveCSS)
            ws.close()
          })

          ws.on('open', () => server.reload('index.js'))
        })

        it('should broadcast to all clients', function (done) {
          let opened = 0
          let reloaded = 0
          let closed = 0

          const ws1 = new WebSocket(base)
          ws1.on('open', open)
          ws1.on('error', (err) => done(err))
          ws1.on('message', message)
          ws1.on('close', close)

          const ws2 = new WebSocket(base)
          ws2.on('open', open)
          ws2.on('error', (err) => done(err))
          ws2.on('message', message)
          ws2.on('close', close)

          function open () {
            if (++opened < 2) return
            server.reload('index.css')
            setTimeout(() => {
              ws1.close()
              ws2.close()
            }, 100)
          }

          function message () {
            reloaded++
          }

          function close () {
            if (++closed < 2) return
            assert.strictEqual(reloaded, 2)
            done()
          }
        })
      })

      context('on server alert', function () {
        it('should send an alert command', function (done) {
          const ws = new WebSocket(base)
          ws.on('close', () => done())
          ws.on('error', (err) => done(err))
          ws.on('message', (payload) => {
            const message = JSON.parse(payload)
            assert.equal(message.command, 'alert')
            assert.equal(message.message, 'hello world')
            ws.close()
          })

          ws.on('open', () => server.alert('hello world'))
        })

        it('should broadcast to all clients', function (done) {
          let opened = 0
          let reloaded = 0
          let closed = 0

          const ws1 = new WebSocket(base)
          ws1.on('open', open)
          ws1.on('error', (err) => done(err))
          ws1.on('message', message)
          ws1.on('close', close)

          const ws2 = new WebSocket(base)
          ws2.on('open', open)
          ws2.on('error', (err) => done(err))
          ws2.on('message', message)
          ws2.on('close', close)

          function open () {
            if (++opened < 2) return
            server.alert('hello world')
            setTimeout(() => {
              ws1.close()
              ws2.close()
            }, 100)
          }

          function message () {
            reloaded++
          }

          function close () {
            if (++closed < 2) return
            assert.strictEqual(reloaded, 2)
            done()
          }
        })
      })
    })
  })
})


lib/websocket.js

'use strict'

const http = require('./http')
const pkg = require('../package.json')
const ws = require('ws')

const port = 35729
const protocols = [
  'http://livereload.com/protocols/connection-check-1',
  'http://livereload.com/protocols/official-7'
]
const serverName = pkg.name

class Server {
  constructor () {
    const wss = new ws.Server({
      server: http(),
      path: '/livereload'
    })

    wss.on('connection', client => this.connect(client))
    this.wss = wss
  }

  connect (client) {
    client.on('message', message => this.handle(client, JSON.parse(message)))
  }

  handle (client, message) {
    if (message.command !== 'hello') return
    this.command(client, 'hello', { protocols })
    // TODO: handle url command?
    // @see http://feedback.livereload.com/knowledgebase/articles/86174-livereload-protocol
  }

  command (client, key, params) {
    const payload = Object.assign({
      command: key,
      serverName: serverName
    }, params)

    client.send(JSON.stringify(payload))
  }

  broadcast (key, params) {
    for (const client of this.wss.clients) {
      this.command(client, key, params)
    }
  }

  reload (file) {
    this.broadcast('reload', {
      path: file,
      liveCSS: true // TODO: make configurable
    })
  }

  alert (message) {
    this.broadcast('alert', { message })
  }

  listen (callback) {
    return this.wss._server.listen(port, callback)
  }

  close (callback) {
    return this.wss.close(callback)
  }
}

module.exports = Server
